pkg/types/query/query.go

```diff
@@ -0,0 +1,294 @@
+package query
+
+import "fmt"
+
+// Visitor should have a per chain per db type implementation that converts primitives to db queries.
+type Visitor interface {
+	ComparerPrimitive(primitive ComparerPrimitive)
+	BlockPrimitive(primitive BlockPrimitive)
+	ConfirmationPrimitive(primitive ConfirmationsPrimitive)
+	TimestampPrimitive(primitive TimestampPrimitive)
+	TxHashPrimitives(primitive TxHashPrimitive)
+}
+
+// Primitive is the basic building block for KeyFilter.
+type Primitive interface {
+	Accept(visitor Visitor)
+}
+
+// KeyFilter is used to filter down chain specific data related to a key.
+type KeyFilter struct {
+	// Key points to the underlying chain contract address and some data that belongs to that contract.
+	// Depending on the underlying Chain Reader blockchain implementation key can map to different onchain concepts, but should be able to map differing onchain data to same offchain data if they belong to the same key.
+	Key string
+	// The base Expressions slice indicates AND logical operation over expressions, which can be primitives or nested boolean expressions.
+	// For eg. []Expression{primitive, primitive, BoolExpression{AND, primitive, BoolExpression{OR, primitive, primitive}} is primitive AND primitive AND (primitive AND (primitive OR primitive)).
+	Expressions []Expression
+}
+
+// Expression contains either a Primitive or a BoolExpression.
+type Expression struct {
+	Primitive      Primitive
+	BoolExpression BoolExpression
+}
+
+func (expr Expression) IsPrimitive() bool {
+	return expr.Primitive != nil
+}
+
+type BoolOperator int
+
+const (
+	AND BoolOperator = iota
+	OR
+)
+
+func (op BoolOperator) String() string {
+	switch op {
+	case AND:
+		return "AND"
+	case OR:
+		return "OR"
+	default:
+		return "Unknown"
+	}
+}
+
+// BoolExpression allows nesting of boolean expressions with different BoolOperator's.
+type BoolExpression struct {
+	// should have minimum length of two
+	Expressions []Expression
+	BoolOperator
+}
+
+func And(expressions ...Expression) Expression {
+	return Expression{
+		BoolExpression: BoolExpression{Expressions: expressions, BoolOperator: AND},
+	}
+}
+
+func Or(expressions ...Expression) Expression {
+	return Expression{
+		BoolExpression: BoolExpression{Expressions: expressions, BoolOperator: OR},
+	}
+}
+
+type ComparisonOperator int
+
+const (
+	Eq ComparisonOperator = iota
+	Neq
+	Gt
+	Lt
+	Gte
+	Lte
+)
+
+type ValueComparer struct {
+	Value    string
+	Operator ComparisonOperator
+}
+
+// ComparerPrimitive is used to filter over values that belong to key data.
+type ComparerPrimitive struct {
+	Name           string
+	ValueComparers []ValueComparer
+}
+
+// Comparer is used for filtering through specific key values.
+// e.g. of filtering for key that belongs to a token transfer by values: Comparer("transferValue", [{"150",LTE}, {"300",GTE}])
+func Comparer(name string, valueComparers ...ValueComparer) Expression {
```

Comment:
Should it technically be s/Comparer/Comparator?
```suggestion
func Comparator(name string, valueComparators ...ValueComparator) Expression {
```
Although an altogether different word might be nice :thinking:

Reply:
Changed to `Comparator`, don't have any fitting naming ideas that are better tbh